### PR TITLE
feat: stack trace style debug print for error

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

### TL;DR

Change the error logging style to

(by `error!` macro)
```
2023-09-25T16:56:56.012879Z ERROR frontend::instance: Failed to execute query: select count(*) from tt trace_id=7112117781994278907 err.msg=0: Table operation error, at src/frontend/src/instance.rs:417:14
1: Failed to plan statement
2: Cannot plan SQL: SELECT count(*) FROM tt, at src/query/src/planner.rs:79:60
3: Plan("Table not found: greptime.public.tt") err.code=PlanQuery
```

(by debug print)
```
0: Table operation error, at src/frontend/src/instance.rs:417:14
1: Failed to plan statement
2: Cannot plan SQL: SELECT count(*) FROM tt, at src/query/src/planner.rs:79:60
3: Plan("Table not found: greptime.public.tt")
```

And the previous one looks like
```
2023-09-25T16:37:56.266934Z ERROR frontend::instance: Failed to execute query: select count(*) from tt trace_id=7112112991553851383 err.msg=Table operation error err.code=PlanQuery err.source=Failed to plan statement err.source.sources=[Cannot plan SQL: SELECT count(*) FROM tt, Error during planning: Table not found: greptime.public.tt]
```

### How to use

Take `cmd/error.rs` as an example:
```diff
-#[derive(Snafu)]
#[snafu(visibility(pub))]
+#[stack_trace_debug]
pub enum Error {
    #[snafu(display("Failed to create default catalog and schema"))]
    InitMetadata {
        location: Location,
        source: common_meta::error::Error,
    },

    #[snafu(display("Cannot create REPL"))]
    ReplCreation {
+        #[snafu(source)]
+        error: ReadlineError,
        location: Location,
    },
}
```
Only a few things to notice:
- need not derive `Debug` manually, the new macro `stack_trace_debug` will generate one that can print the error prettier
- add `#[stack_trace_debug]` to every error enum
- For external errors, change its name to `error` and add a snafu annotation `#[snafu(source)]`.

### todo
- remove `location_opt`
- maybe add a method to retrieve external error source

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- https://github.com/GreptimeTeam/greptimedb/issues/1854

### CODE REVIEW

File: src/client/examples/stream_ingest.rs [hunk 1]
```diff
@@ -42,14 +42,14 @@ async fn run() {
         .insert(vec![to_insert_request(weather_records_1())])
         .await
     {
-        error!("Error: {e}");
+        error!("Error: {e:?}");
```
[the remark this answers is not in the document]
Comment (waynexia) [reply src/client/examples/stream_ingest.rs:45]: I've changed the `error!` and `warn!` to display debug output by default. Error's display is generated by snafu, which doesn't have enough information. So I suppose there won't be a situation in which we need to use display directly?

**Review by shuiyisong — APPROVED** [review]
LGTM

File: src/client/examples/stream_ingest.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (evenyag) [reply src/client/examples/stream_ingest.rs:45]: > I've changed the `error!` and `warn!` to display debug output by default.

This only works for errors that implement `ErrorExt`. If we support `error!(?e; "xxx")` then we could use it to print all other errors that don't implement `ErrorExt`.

**Review by evenyag — APPROVED** [review]
Since this PR is easy to conflict with others, I decided to merge it first.

We could implement the enhancement mentioned in https://github.com/GreptimeTeam/greptimedb/pull/2489#discussion_r1336571235 later
